### DISCUSSION TITLE
Thoughts on Implementing ISR like Next JS

### DISCUSSION THREAD

**Discussion by The-Real-Thisas** [withastro/roadmap#228] [Proposal]
I think for real time news sites and similar that want the performance of SSG without having to rebuild the entire site on every update, this is incredible.

I don't see support in the roadmap but curious if it's on the team's radar by any chance ? 

[https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration
)

**Comment by fschroiff** [comment]
And please make it compatible with all Hosts 🥺

**Comment by FredKSchott** [comment]
Astro currently has a feature called "hybrid rendering" which lets you mix static, pre-rendered content with dynamic rendered-on-demand content.

You can configure the length of time that a response is cached in your CDN via the `Cache-Control` header. Astro doesn't currently have a built-in `invalidate` API for this, but we could add one if there was interest!

**Reply by wassfila** [reply]
dynamic rerender does not work for e.g. mdx, which still need to be static.

**Reply by omerman** [reply]
There is interest.

**Reply by eric-burel** [reply]
For the record, I've explored the SSR (on-demand rendering in Astro wording, dynamic rendering in Next wording, per-request server-side rendering in my own non ambiguous wording :)) + HTTP cache approach a bit: https://blog.vulcanjs.org/treat-your-users-right-with-http-cache-and-segmented-rendering-7a4f4761b549
It's able to mimick static rendering in most useful scenarios, however it should be noted that it falls short in a few specific situations: 
- using basic auth (because of the authorization header)
- if you rely on URL rewrites to personalize content (advanced use case but still happens eg for AB tests)

For these a true ISR = caching at the application layer rather than relying on the CDN layer with Cache-control is more suited.

**Comment by AkashRajpurohit** [comment]
Would recommend this video for anyone who wants to deep dive into understanding the different cache headers and how ISR can be implemented via CDN + cache headers

https://www.youtube.com/watch?v=bfLFHp7Sbkg

**Comment by lilnasy** [comment]
ISR is going to require different approaches that depend on where you are deploying. I am working on a guide for the docs site that you can track here: withastro/docs#3527 ([rendered preview](https://docs-git-fork-lilnasy-main-astrodotbuild.vercel.app/en/guides/incremental-static-regeneration/)).

Please share how you plan to use ISR and where you deploy so I can better document it.

**Reply by ohrrkan** [reply]
@lilnasy, after reading the linked doc, wondering if there is not a small confusion between cache control and ISR? They are some few differences between both.  Astro do not support (for now) ISR.

The docs is amazing but I will rename it to cache-control to avoid confusion before you merge. By the way really vote for Astro to implement ISR in futur.

**Reply by lilnasy** [reply]
You are right, they are not the same thing.

I think the one difference would be that ISR entails the CDN _not_ evicting the page once built. This is already supported by the netlify adapter and documented in the doc. Cloudflare does not offer a way to make that happen. I haven't looked into what implementing it for Vercel would look like.

**Reply by ohrrkan** [reply]
Concerning Vercel you can find the diffs in this table https://vercel.com/docs/incremental-static-regeneration#isr-vs-cache-control-comparison-table

**Reply by lilnasy** [reply]
I should clarify. What I meant was, I could not find their docs for framework developers hoping to implement ISR when deploying to Vercel. I'm pretty sure it can be done with our vercel adapter.

**Reply by aretrace** [reply]
Your documentation is very helpful, hopefully it can be integrated into the main documentation soon.

**Reply by lilnasy** [reply]
I'm pretty sure it will be, although it will be framed around "scalability" rather than one specific feature that enables scalability.

**Comment by lilnasy** [comment]
The vercel adapter now supports Vercel's implementation of ISR. [@astrojs/vercel | Astro Docs](https://docs.astro.build/en/guides/integrations-guide/vercel/#isr).

Netlify's equivalent is called [On-Demand Builders](https://docs.netlify.com/configure-builds/on-demand-builders), and is usable in version 3.1.1 of the official adapter. The current iteration (v4, v5) offers [automatic caching](https://docs.astro.build/en/guides/integrations-guide/netlify/#caching-pages) that uses standard `Cache-Control` headers under the hood, instead. The difference from builders is that the cache is per-region, and temporary, instead of global and persistent.

Cloudflare has the infrastructure to offer an equivalent in [Cache Reserve](https://www.cloudflare.com/products/cache-reserve/), but it is not hooked up to the imperative [Cache API](https://developers.cloudflare.com/workers/runtime-apis/cache/) available in workers. Your only option is per-region and temporary caching.

For self-hosting, the specifics will depend on the exact set of reverse proxies you are using, but the idea will be to make them save cached responses in object storage.

**Comment by bholmesdev** [comment]
Looks like this issue has been resolved! We now support adapter-level config for ISR on major platforms that support it, notably Vercel and Netlify. [Try it out](https://docs.astro.build/en/guides/integrations-guide/vercel/#isr) and feel free to open further discussion for features you'd like to see.

**Comment by monaye** [comment]
I am curious, if both [@astrojs/vercel | Astro Docs](https://docs.astro.build/en/guides/integrations-guide/vercel/#isr) and [On-Demand Builders](https://docs.netlify.com/configure-builds/on-demand-builders) will work with the [Server Island](https://docs.astro.build/en/guides/server-islands/) ?? if so it would nice to documented (I can if someone can confirm).

**Comment by heggemsnes** [comment]
Awesome that Astro is getting ISR! This seems like the final steps from leaving frustrating Next.js for Astro, but on demand revalidation when? :D Mostly via cache tags would be great.

**Reply by FilipLjubic** [reply]
https://vercel.com/docs/build-output-api/features#on-demand-incremental-static-regeneration-isr

**Comment by maxn0d3x** [comment]
We need ISR

**Reply by delucis** [reply]
ISR is a Vercel feature and is supported for Astro with the Vercel adapter: https://docs.astro.build/en/guides/integrations-guide/vercel/#isr

A similar Netlify feature is also available: https://docs.astro.build/en/guides/integrations-guide/netlify/#caching-pages

For other platforms you can achieve similar things using CDN cache headers. See for example:

- https://logsnag.com/blog/implementing-isr-in-astro
- https://developers.netlify.com/guides/how-to-do-advanced-caching-and-isr-with-astro/

**Reply by s4093120** [reply]
What if the content rarely changes but you keep creating new content? Setting a cache header doesn't seem the best as 1. setting short cache headers causes excessive render requests and 2. setting extremely long headers prevents content from being updated quickly. It would be great if we could invalidate certain pages when it changes.

**Reply by delucis** [reply]
@rjawaji1 The “cache tag” approach described in the Netlify blog post I linked above is one common approach, which allows for long cache times which can be invalidated when content changes.

**Reply by kelenakamura** [reply]
is there a vercel equivalent?

**Reply by delucis** [reply]
> ISR is a Vercel feature and is supported for Astro with the Vercel adapter: [docs.astro.build/en/guides/integrations-guide/vercel#isr](https://docs.astro.build/en/guides/integrations-guide/vercel/#isr)

**Reply by kelenakamura** [reply]
i meant the on-demand ISR, netlify doc uses cache-tags, can you also do that with vercel? i know with nextjs you use their revalidateTag method. not sure on the equivalent Astro way.

**Reply by delucis** [reply]
Oh right. I think you can do it the same way with Cache-Control headers? https://vercel.com/docs/edge-cache

**Reply by kelenakamura** [reply]
actually found this post - https://www.shawn.so/blog/isr-astro-vercel-on-demand-incremental-static-regeneration/ for route based on-demand ISR. not tag based though.
